# Chapter 4: Functions

[**4.1 Functions**](#4.1-Functions)   
[**4.2 User Defined Functions**](#4.2-User-Defined-Functions)   
[**4.2.1 Define Functions**](#4.2.1-Define-Functions)   
[**4.2.2 Variable Scope**](#4.2.2-Variable-Scope)   
[**4.2.3 Function Documentation**](#4.2.3-Function-Documentation)   
[**4.2.4 Function Parameter**](#4.2.4-Function-Parameter)   
[**4.2.5 Default Parameter Values**](#4.2.5-Default-Parameter-Values)   
[**4.2.6 Keyword Arguments**](#4.2.6-Keyword-Arguments)   
[**4.2.7 Arbitrary Arguments Lists**](#4.2.7-Arbitrary-Arguments-Lists)   
[**4.3 Recursion**](#4.3-Recursion)  
[**4.4 Module**](#4.4-Module)  
[**4.5 Lambda**](#4.5-Lambda)    
[**Data Analytics**](#Data-Analytics)   

#### 4.1 Functions
**Functions**: Functions is a set of statements that perform certain task based on the input and generate output or modify values. Functions are created to organize the code that repeat multiple times. When there is a repeated code we can organize them creating a function and call them where ever we need. Function helps to reduce the amount of repeated code, organize the code based on business logic, maintain code easily for refactoring or implementing changes.  
We have already discussed many built-in functions such as:  
`print`  
`type`  
`input`  
`datetime`  
`range`  
`random` and several `string functions`.  
*Note: Python has many useful built-in functions so before writing your own function check if functions exit which helps to eliminate redundancy.*

#### 4.2 User Defined Functions
User Defined Functions (UDF) also known as custom functions are created to break the large program into chunk of  small logical unit to solve specific business problem. UDF are created not only to solve problem but also to faciliate functionality of program. i.e. to support the functions itself such as check file types, generate custom timestamp, calculate total job runtime etc.

#### 4.2.1 Define Functions
A function is defined with `def` keyword followed by `function-name`, list of parameter surrounded by parentheses `()` and a colon `:`. The function name should always begin with a lowercase letter and if it contains multiple words then it should use underscore `_` to seperate the word.  
*Note: The recommended parameter list should always be as minimum as possible to reduce complexity and unit test.*  

**Parameter** is a variable defined in method definition.   
**Argument** is the actual value which is passed while calling a function.  

The line after `:` consists of function's block with an optional docstring in first line. Docstring explains the purpose of function defined on multiple line. It is very useful to provide docstring which helps other developer to understand and easily maintain/modify the code in future. It is also useful for documentation purpose. If the function produce output then it should have `return` statement. The `return` statement terminate the function and generate the output to the caller.

In [1]:
def convert_fahrenheit_celsius(degree_fahrenheit):
    """Convert degree Fahrenheit to degree Celsius
    input: degree_fahrenheit
    output: degree_celsius    
    """
    celsius = (degree_fahrenheit - 32) * (5/9)
    rounded_celsius = round(celsius,2)
    return rounded_celsius

degree_celsius = convert_fahrenheit_celsius(-19)
print(degree_celsius)

-28.33


The code snippet above described briefly:
* Line 1: `def` keyword defines the begining of a function, `convert_fahrenheit_celsius` is a function name, `degree_fahrenheit` is input parameter to a function, and `:` is terminate the function header and starts the function blocks. 
* Line 2-5: docstring inside triple quote `"""`.
* Line 6: `celsius` local variable to store the calculation's value for `(degree_fahrenheit - 32) * (5/9)`
* Line 7: `rounded_celsius` local variable to store the 2 digit round value from `round(celsius,2)`
* Line 8: `return` keyword to output the `rounded_celsius` value from the function.
* Line 10: `degree_celsius`  is a variable to store the value return by function `convert_fahrenheit_celsius`. **-19** value is argument to function. The output value return from function is **-28.33**.

#### 4.2.2 Variable Scope
**Local Variable**: All the variables defined inside the function are local variables. The scope of function is valid only inside the function but outside the function. You'll get `NameError` when you try to access the variable outside the functions shown below.

In [2]:
from datetime import datetime
def get_date():
    """
    Get current date with string type in yyyy-mm-dd format.
        >>> get_date()
        2019-08-28
    :return: current date  
    :rtype: string
    """
    curr_timestamp = datetime.now()
    curr_date = curr_timestamp.strftime("%Y-%m-%d")
    return curr_date

In [3]:
print("Today is %s" % get_date())
print("Today is %s" % curr_timestamp) # Accessing the local variable outside a functions

Today is 2019-09-08


NameError: name 'curr_timestamp' is not defined

In [4]:
!ls # list directory
!mkdir {get_date()} # make directory with current date by calling get_date function.
                    # {varname} is used to pass the python variables into the shell.
!ls # list directory

Chapter_4.ipynb       data_analysis_3.ipynb
2019-09-08            Chapter_4.ipynb       data_analysis_3.ipynb


**Global Variable**: All the variables defined outside any function or class are known as `global variables`. Global variables can be used anywhere inside a python file or interactive session after they are declared. The value of global variable cannot be changed inside a function. The scope of functions will be local only inside the function.

In [5]:
value = 50 # global variable
def variable_scope():
    value = 10 # local variable
    print("Value inside fuction:",value)
    
print("Before function call:",value)
variable_scope()
print("After function call:",value)

Before function call: 50
Value inside fuction: 10
After function call: 50


In order to modify the value of global variable inside the function `global` stamtent should be declared inside the function.

In [6]:
value = 50 # global variable
def variable_scope():
    global value
    value = 10 # local variable
    print("Value inside fuction:",value)
    
print("Before function call:",value)
variable_scope()
print("After function call:",value)

Before function call: 50
Value inside fuction: 10
After function call: 10


#### 4.2.3 Function Documentation
The docstring is used to provide description about function. We can access the docstring by passing function name followed by `?`. `??` is used to display all the source code of functions. `help` is also used to describe the functions description. You can choose which your favourite method to display the help information. I prefer to use `help`.

In [7]:
help(get_date)  #describe function 
get_date?
get_date??

Help on function get_date in module __main__:

get_date()
    Get current date with string type in yyyy-mm-dd format.
        >>> get_date()
        2019-08-28
    :return: current date  
    :rtype: string



#### 4.2.4 Function Parameter
Function can contain several parameters. The parameter to a function is choosen based on the business requirement/logic. The parameter can be any types i.e. int, float, boolean, string, object etc.

In [8]:
def tokenizer(sentence, stats):
    string_len = len(sentence)
    string_token = sentence.split()
    print("Input String:%s" % sentence)    
    if(stats):
        import pandas as pd
        print("=================\n   Statistics\n=================")
        print("String length: %s." % string_len)
        df_string = pd.Series(string_token)
        df_count = df_string.groupby(df_string).size()
        print(df_count)    

tokenizer("Hello this is a a new class. Hellow this is a a new class. Hellow this is a a new class.", True)

Input String:Hello this is a a new class. Hellow this is a a new class. Hellow this is a a new class.
   Statistics
String length: 88.
Hello     1
Hellow    2
a         6
class.    3
is        3
new       3
this      3
dtype: int64


#### 4.2.5 Default Parameter Values
Values which are defined while creating functions are known as `default parameter values`. When a function is created with default paramater values then while calling the function without any parameter the default value will be passed. But if we pass the parameter then the default value will be ignored. Default parameter values is set by **parameter=value**. Arguments are always assigned from left to right. The first value will always leftmost argument value.

In [9]:
def greeting(name='John', message='Welcome', ):
    greet = message + " " + name + "."
    return greet

greeting() # no argument

'Welcome John.'

In [10]:
print(greeting("Python")) # one argument
print(greeting("Pandas", "Learning")) # two argument
print(greeting("Pandas", "Learning", "is good")) # what happen if we pass three argument

Welcome Python.
Learning Pandas.


TypeError: greeting() takes from 0 to 2 positional arguments but 3 were given

#### 4.2.6 Keyword Arguments
`Keyword arguments` is used to pass arguments in **any** order. Keyworkd argument uses `parameter_name=value` form where parameter_name is the actual name of parameter and value is associated values for that parameter. If the parameter name is not specified then by default the value are chosen from left to right. Keyword arguments helps make code readability in function calls where function has many arguments.

In [11]:
def arithmetic_operation(first_value, second_value, operation):
    if(operation == 'add'):
        calculation = first_value + second_value
    elif(operation == 'subtract'):
        calculation = first_value - second_value
    elif(operation == 'multiply'):
        calculation = first_value * second_value
    elif(operation == 'divide'):
        calculation = first_value / second_value
    elif(operation == 'modulo'):
        calculation = first_value % second_value
    elif(operation == 'exponent'):
        calculation = first_value ** second_value
    else:
        print("Undefined operation:{}. Returns None value".format(operation))
        return None
    return calculation

arithmetic_operation(first_value=5, second_value=10, operation='add')

15

In [12]:
print(arithmetic_operation(second_value=50, first_value=20, operation='subtract'))
print(arithmetic_operation( operation='divide',first_value=50, second_value=20))
print(arithmetic_operation(2,10,'exponent'))

-30
2.5
1024


#### 4.2.7 Arbitrary Arguments Lists
`Arbitrary arguments lists` functions can have any number of arguments. The parameter is defned with `*args` form. The **\*** before **args** indicate Python to accept any number of parameter to be packed in the form of tuples. The `args` parameter name is used by convention but we can use any name instead. If there are multiple parameters then **\*args** must be always rightmost parameter in function.

Looking at the Python built-in function such as `min`, `max` etc. Those function requires two values i.e. *arg1* and *arg2* and optional third parameter *\*args*.

In [13]:
help(min)

Help on built-in function min in module builtins:

min(...)
    min(iterable, *[, default=obj, key=func]) -> value
    min(arg1, arg2, *args, *[, key=func]) -> value
    
    With a single iterable argument, return its smallest item. The
    default keyword-only argument specifies an object to return if
    the provided iterable is empty.
    With two or more arguments, return the smallest argument.



In [14]:
print(min(10,50))
print(min(10,50,100,900,1))
print(min(0,10,50,0,4,3,2))

10
1
0


In [15]:
def mean(*args):
    return sum(args) / len(args)

In [16]:
print(mean(10))
print(mean(10,20))
print(mean(10,20,20,5,6,10))

10.0
15.0
11.833333333333334


**Passing Iterable Elements**: The values of tuple, list and other iterable elements can be pass into function as individual arguments by using `* operator`. This operator unpacks all the element as individual parameter into a function as an arguments.

In [17]:
score = [10,20,30,40,50,20,10,20,40,10]
print(mean(*score)) # same as mean(10,20,30,40,50,20,10,20,40,10)

25.0


**\*\*kwargs**: It is known as keyword arguments which is used to pass keyworded argument list. It uses `**` instead of `*` . In keyword argument, the key/value is passed as function argument. It also accept any numbers of argument similar to `*args`. Dictionary can also be passed in `kwargs`.

In [18]:
def countries_abbr(**kwargs):
    print(kwargs)
    
countries_abbr(us="United States",uk="United Kingdom")

{'us': 'United States', 'uk': 'United Kingdom'}


In [19]:
def print_information(**kwargs):
    for k, v in kwargs.items():
        if k == 'name':
            print("Full Name:",v)
        if k == 'course':
            print("Course Name:",v)
        if k == 'grade':
            print("Course Grade:",v)
            
            
student_info={'name':'John Doe', 'course':'Python', 'grade':9.5}
print_information(**student_info)   

Full Name: John Doe
Course Name: Python
Course Grade: 9.5


**Reading Assignment**
* Pass-by-value
* Pass-by-reference

#### 4.3 Recursion
When a function calls itself directly or indirectly then it is known as `recursion`. Those types of fuction is called `recursive function`. Recursive algorithm helps to easily solve complex problem. Various problem solved with recursion includes Towers of Hanoi, Tree Traversal etc. **Stack overflow** occurs if there is infinite recursion.

In [20]:
def factorial(value):
    if value <=1:
        return 1
    return value * factorial(value - 1)

factorial_value = factorial(5)
print("Factorial of 5:",factorial_value)   

Factorial of 5: 120


#### 4.4 Module
The Python file that contains Python definitions and statements is known as `module`. A module helps to organize code. In module, we can define functions, classes and variables. The file name for module is module name with suffix `.py`. The module name is available in global variable `__name__`.

**Importing Module**: 
* A module can imported in the Python code with various ways:
* `import modulename` allows to import specific module. 
* `import modulename as mod` allows to bound all the modulename to mod.  
* `from modulename import name1..nameN` allows to import specific attributes (provided in names) from a module.  
* `from modulename import *` allows to import all names from a module.  

In [24]:
from analyticstensor import word_count
my_string = "Hey I want to count how many words we have in this sentence, You can try this module to test."
wc = word_count(my_string)
wc

Input String:Hey I want to count how many words we have in this sentence, You can try this module to test.
   Statistics
Total words: 20.


this         2
to           2
words        1
want         1
You          1
in           1
how          1
try          1
we           1
have         1
test.        1
I            1
count        1
Hey          1
module       1
sentence,    1
many         1
can          1
dtype: int64

#### 4.5 Lambda
Anonymous function is a function without no names. `lambda` keyword is used to create anonymous functions in Python.  
The syntax of lambda function is  
`lambda arguments: expression`  
Lambda function is an anonymous, single-use which can be created in one line of code. These lambda expression can have any numbers of argument but only one expression. It cannot contain any statements. It doesn't return function object for assigning to the variable. The expression is always return value.

In [22]:
add = lambda x, y: x + y
add(10, 15)

25

In [23]:
square = lambda x: x * x
square(5)

25

**Lambda are mostly used in sorting, filtering and map functions.**

**Example 1: Sorting a list elements**

In [25]:
course = ["MySQL", "Python", "Spark", "Java", "Machine Learing"]
course.sort(key = lambda x: x)
print(course)

numbers = [100, 49, 58, 0, 10, 50]
numbers.sort(key = lambda x: x)
print(numbers)

['Java', 'Machine Learing', 'MySQL', 'Python', 'Spark']
[0, 10, 49, 50, 58, 100]


**Example 2: Sorting a list of tuples**

In [26]:
grocery = [('apple', 5.25), ('cabbage', 2.34), ('orange', 3.00), ('biscuit', 5.54), ('pen', 2.39), ('milk', 2.49)]
grocery.sort(key = lambda x: x[0]) # sort by item name
print(grocery)
grocery.sort(key = lambda x: x[1]) # sort by item price
print(grocery)

[('apple', 5.25), ('biscuit', 5.54), ('cabbage', 2.34), ('milk', 2.49), ('orange', 3.0), ('pen', 2.39)]
[('cabbage', 2.34), ('pen', 2.39), ('milk', 2.49), ('orange', 3.0), ('apple', 5.25), ('biscuit', 5.54)]


**Example 3: Sorting a list of dictionaries**

In [27]:
import pprint as pp
student = [{'name':'David', 'age':25, 'enrolled': 2018}, {'name':'John', 'age':24, 'enrolled': 2019}, 
          {'name':'Alice', 'age':26, 'enrolled': 2016}]
student_sorted = sorted(student, key = lambda x: x['enrolled'])
pp.pprint(student_sorted)

[{'age': 26, 'enrolled': 2016, 'name': 'Alice'},
 {'age': 25, 'enrolled': 2018, 'name': 'David'},
 {'age': 24, 'enrolled': 2019, 'name': 'John'}]


**Example 4: Filtering a list of integers**

In [28]:
num = [ 10, 15, 19, 45, 20, 100, 69]
num1 = list(filter(lambda x: x%2 == 0, num))
num1

[10, 20, 100]

In [29]:
odd = lambda x: x%2 == 1
num = [ 10, 15, 19, 45, 20, 100, 69]
num1 = list(filter(odd, num))
num1

[15, 19, 45, 69]

**Example 5: Map function**

In [30]:
num = [ 10, 15, 19, 45, 20, 100, 69]
num1 = list(map(lambda x: x ** 2, num))
num1

[100, 225, 361, 2025, 400, 10000, 4761]

**Lambda with conditionals**  
`lambda args: a if boolean_expression else b`

In [31]:
begin_with_S = lambda x: True if x.startswith('S') else False
print(begin_with_S('South'))

True
